x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutDataFrameAnalyticsAction.java

```diff
@@ -46,12 +71,60 @@ protected void doExecute(Task task, PutDataFrameAnalyticsAction.Request request,
-        ));
+        if (licenseState.isAuthAllowed()) {
+            final String username = securityContext.getUser().principal();
+            RoleDescriptor.IndicesPrivileges sourceIndexPrivileges = RoleDescriptor.IndicesPrivileges.builder()
+                .indices(request.getConfig().getSource())
+                .privileges(SearchAction.NAME)
+                .build();
+            RoleDescriptor.IndicesPrivileges destIndexPrivileges = RoleDescriptor.IndicesPrivileges.builder()
+                .indices(request.getConfig().getDest())
+                 // CreateIndexAction.NAME is more expansive than and create_index, in this instance we want at least create_index.
+                .privileges(SearchAction.NAME, IndexAction.NAME, "create_index")
+                .build();
+
+            HasPrivilegesRequest privRequest = new HasPrivilegesRequest();
+            privRequest.applicationPrivileges(new RoleDescriptor.ApplicationResourcePrivileges[0]);
+            privRequest.username(username);
+            privRequest.clusterPrivileges(Strings.EMPTY_ARRAY);
+            privRequest.indexPrivileges(sourceIndexPrivileges, destIndexPrivileges);
+
+            ActionListener<HasPrivilegesResponse> privResponseListener = ActionListener.wrap(
+                r -> handlePrivsResponse(username, request, r, listener),
+                listener::onFailure);
+
+            client.execute(HasPrivilegesAction.INSTANCE, privRequest, privResponseListener);
+        } else {
+            configProvider.put(request.getConfig(), threadPool.getThreadContext().getHeaders(), ActionListener.wrap(
+                indexResponse -> listener.onResponse(new PutDataFrameAnalyticsAction.Response(request.getConfig())),
+                listener::onFailure
+            ));
+        }
+    }
+
+    private void handlePrivsResponse(String username, PutDataFrameAnalyticsAction.Request request,
+                                     HasPrivilegesResponse response,
+                                     ActionListener<PutDataFrameAnalyticsAction.Response> listener) throws IOException {
+        if (response.isCompleteMatch()) {
+            configProvider.put(request.getConfig(), threadPool.getThreadContext().getHeaders(), ActionListener.wrap(
+                indexResponse -> listener.onResponse(new PutDataFrameAnalyticsAction.Response(request.getConfig())),
+                listener::onFailure
+            ));
+        } else {
+            XContentBuilder builder = JsonXContent.contentBuilder();
+            builder.startObject();
+            for (ResourcePrivileges index : response.getIndexPrivileges()) {
+                builder.field(index.getResource());
+                builder.map(index.getPrivileges());
+            }
+            builder.endObject();
+
+            listener.onFailure(Exceptions.authorizationError("Cannot create data frame analytics [{}]" +
+                    " because user {} lacks permissions on the indices: {}",
```

Comment:
Is it possible to say which permissions are missing?

Reply:
The error message includes them in the body. It shows what permissions are `true` and which are `false`.